### ISSUE TITLE
Create VirtualTypes for Contollers using di.xml

### ISSUE THREAD

**Issue by qsolutions-pl** [magento/magento2#18151]
### Description

Make Magneto 2 even more flexible then it is now :-)
We can create virtual types for blocks, models & most of the classes used in the system.
I think it would be great if we could use VirtualTypes for Controllers.

### Expected behavior

Use a custom class for an action that does not exists in CORE modules simply by adding 1 node in **di.xml** file

### Benefits

We can extend core routers with VirtualTypes. For example: I need to add new action to a CORE controller, I can simply do it like this
`<virtualType name="Magento\CatalogSearch\Controller\Result\Somerandomaction" type="Qextensions\NoResults\Controller\Mycustomaction" />`

### Additional information

I hope my description is selfexplanatory

**Comment by magento-engcom-team** [comment]
Hi @qsolutions-pl. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@qsolutions-pl do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `G1 Passed` will be added to the issue automatically. Please, edit issue description if needed, until label `G1 Passed` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `G2 Passed` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `acknowledged` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report is acknowledged.

**Comment by pmclain** [comment]
A little background on why this doesn't work at the moment. The core routers build the action list using `\Magento\Framework\Module\Dir\Reader::getActionFiles` source, which loads the available actions by reading files contained in each module's Controller directory.
https://github.com/magento/magento2/blob/5916c5a0c3eb595ace91448c35a170924cd1c4e1/lib/internal/Magento/Framework/Module/Dir/Reader.php#L123-L136

**Comment by jakwinkler** [comment]
@pmclain right, but I think extending virtual types would be awesome to create extra action to core controllers.

**Comment by orlangur** [comment]
@qsolutions-pl,

> create extra action to core controllers

How is that different from URL rewrites?

How is line line in `di.xml` easier than an empty controller class in your module? With such change one would have to check not just `Controller` folder but also DI configuration to find a corresponding declaration.
